[WIP] ENH: Added nd_freeze script to allow access to NeuroDebian snapshot r…

mjtravers opened the pull request:
…epositories

For now, the script "knocks" on the neuro.debian.net URL to gain access to the snapshots repo server.

The commit addresses Issue #38 

 So the outstanding TODOs are

- [x] added APT line(s) for `data` suite of neurodebian which were missing
- [x] original APT sources must be disabled whenever snapshot one is provided for that corresponding source (optional: you might like add an option `--keep-original-apt-sources` to disable such removal)
- [x] For Ubuntu itself there is no snapshots, so those APT sources must not be replaced
- [x] Should get a command line option like `--allow-original-apt-sources`, without which one all those for which any original, non-snapshot APT source left, by default we would exit with non-0 exit status.  This is to preclude (by default) a false assumption that we did freeze it.  With any original APT source in place, no freezing is done pretty much.
- [x] Since snapshots repositories might be considered "outdated", we need to adjust apt configuration with smth like `echo 'Acquire::Check-Valid-Until "0";' > /etc/apt/apt.conf.d/10no--check-valid-until`
- [x] To guarantee that we actually have proper access to the APT repositories, we should run `apt-get update` with `--no-allow-insecure-repositories`
- [x] deb-src lines should be added only if they were also for those APT repos in original apt configuration

Probably in the next iteration, after this PR is ready, would be needed to address that usecase I stated

- [ ] If no specific date is specified - replace APT sources to snapshot ones for the  Release date as depicted in their Release files on the system.  So similar to what we were parsing out in NICEMAN.
- [ ] (optionally) upon configuration of all the APT snapshot sources, try to determine which packages are ahead of their time and try to downgrade by providing exact versions as available from the snapshots APT sources.  Would be needed if e.g. base image changes forward but I do want to replicate 100% (or as close as I could since some config changes which wouldn't happen could preclude)

yarikoptic commented on tools/nd_freeze:
```diff
@@ -0,0 +1,167 @@
+done
+
+echo NeuroDebian snapshot repo source information saved to $APT_SOURCE_PATH
+echo Run \"sudo apt-get update\" to make the snapshot packages available.
```
it is ok/safe to run apt-get update right here IMHO.  it was not run within that setup script I pointed you to because that script itself was ran within/by apt process, so we could not invoke `apt-get update` due to locking.  here it should be ok IMHO

yarikoptic commented:
@mjtravers would be nice to finalize this one - I have used it already on a couple of occasions within Dockerfile with great results ;)  Next would be to use it within our [NeuroDebian singularity file](https://github.com/neurodebian/neurodebian/blob/master/singularity/Singularity).

 So the outstanding TODOs are

- [ ] original APT sources must be disabled whenever snapshot one is provided for that corresponding source (optional: you might like add an option `--keep-original-apt-sources` to disable such removal)
- [ ] For Ubuntu itself there is no snapshots, so those APT sources must not be replaced
- [ ] Should get a command line option like `--allow-original-apt-sources`, without which one all those for which any original, non-snapshot APT source left, by default we would exit with non-0 exit status.  This is to preclude (by default) a false assumption that we did freeze it.  With any original APT source in place, no freezing is done pretty much.

Probably in the next iteration, after this PR is ready, would be needed to address that usecase I stated

- [ ] If no specific date is specified - replace APT sources to snapshot ones for the  Release date as depicted in their Release files on the system.  So similar to what we were parsing out in NICEMAN.

yarikoptic commented on tools/nd_freeze:
```diff
@@ -0,0 +1,167 @@
+#   yyyy-mm-dd
+#   yyyymmddThh:mm
+#   "yyyy-mm-dd hh:mm:ss"
+#   "mm/dd/yy hh:mm:ss"
```
are the times in the last types of entry in UTC or local zone?  I guess in the topmost it is the UTC?

mjtravers replied:
The timezone of importance is the one at snapshot-neuro.debian.net. I assume it is UTC so the user should be notified that the date/time they enter is not translated into UTC.

yarikoptic commented on tools/nd_freeze:
```diff
@@ -0,0 +1,167 @@
+
+
+# Check for existence of required command line utilities
+programs='wget sed'
```
Please make it possible to use curl instead of wget (ideally could try both but I guess would be trickier so probably can just switch to use curl).  I found curl more frequently present than wget, e.g. in current stock Debian:latest docker image.
I will push a fix for that since need to get it going

yarikoptic replied:
eh that sucks:
```shell
$> docker run -it --rm -v ~/dicom-test:/dicoms --entrypoint=bash debian:sid
root@79268d5ebfef:/# wget
bash: wget: command not found
root@79268d5ebfef:/# curl
bash: curl: command not found
```
I wonder if there is another "internet" aware utility/thing for basic http transaction likely to reside on such basic/stripped instance...  there is perl...

yarikoptic replied:
ok, here is my patch for `curl` just in case 
```patch
$> git diff
diff --git a/tools/nd_freeze b/tools/nd_freeze
index cd5b7ec..100b0b7 100755
--- a/tools/nd_freeze
+++ b/tools/nd_freeze
@@ -55,7 +55,7 @@ user_date=$1
 
 
 # Check for existence of required command line utilities
-programs='wget sed'
+programs='curl sed'
 for program in $programs
 do
   if [ -z `which ${program}` ]; then
@@ -133,12 +133,12 @@ fi
 
 # Knock on snapshot's door. This is temporarily necessary until the production
 # version of the site is made available.
-wget -O /dev/null -q http://neuro.debian.net/_files/knock-snapshots
+curl -s http://neuro.debian.net/_files/knock-snapshots >| /dev/null
 
 
 # Add "next" snapshot to cover package updates made after the previous snapshot
 # but before the user provided date.
-next_line=`wget -qO- http://snapshot-neuro.debian.net:5002/archive/neurodebian/${previous_timestamp}/ | sed -n '/>next<\/a>/p'`
+next_line=`curl -s http://snapshot-neuro.debian.net:5002/archive/neurodebian/${previous_timestamp}/ | sed -n '/>next<\/a>/p'`
 next_timestamp=`echo "$next_line" | sed -E 's/^.*\/archive\/neurodebian\/([0-9TZ]+)\/">next.*$/\1/'`
 if [ -z "$next_timestamp" ]; then
```
to install `curl` in the stock docker image ATM I need
```
5 upgraded, 26 newly installed, 0 to remove and 82 not upgraded.
Need to get 10.5 MB of archives.
After this operation, 18.3 MB of additional disk space will be used.
```
to install `wget`
```
2 upgraded, 11 newly installed, 0 to remove and 85 not upgraded.
Need to get 6956 kB of archives.
After this operation, 12.4 MB of additional disk space will be used.
```
so, fixes should

- [ ] provide support for either having `wget` or `curl` installed
- [ ] if none of those two is available, should `apt-get install  --no-install-recommends wget` since altogether seems to be leaner.  

NB since perl is there, there might be a way or just a single package to install to get some basic "fetch from the web" functionality working... just dunno one yet

mjtravers commented on tools/nd_freeze:
```diff
@@ -0,0 +1,167 @@
+fi
+
+# Discover distribution codename
+source_line=`cat /etc/apt/sources.list | sed -n '/^deb.*main/p' | head -1`
```
This should work although I am not able to get the release name using apt policy in debian:jessie. Maybe I'll use the /etc/apt/sources.list as a plan B

yarikoptic commented:
a possibly useful bash snippet to "parse" apt .list files
```bash
grep -n '^deb\(-src\|\) ' /etc/apt/sources.list{,.d/*.list} | sed -e 's,list:,list ,g' -e 's,:deb, deb,g' | while read aptfile lineno debsrc url release suites ; do echo $debsrc $url $release $suites; done
```

yarikoptic commented:
grr... couldn't reply for some reason to your comment

   This should work although I am not able to get the release name using apt                                    
   policy in debian:jessie. Maybe I'll use the /etc/apt/sources.list as a                                       
   plan B         

and clicked "Resolve" so it was gone.  I wanted to ask "what do you mean?" ;):
```shell
$> docker run -it --rm debian:jessie
root@312ef70cd34a:/# apt-get update -qqq
root@312ef70cd34a:/# apt-cache policy
Package files:
 100 /var/lib/dpkg/status
     release a=now
 500 http://security.debian.org/ jessie/updates/main amd64 Packages
     release v=8,o=Debian,a=oldstable,n=jessie,l=Debian-Security,c=main
     origin security.debian.org
 500 http://httpredir.debian.org/debian/ jessie-updates/main amd64 Packages
     release o=Debian,a=oldstable-updates,n=jessie-updates,l=Debian,c=main
     origin httpredir.debian.org
 500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
     release v=8.11,o=Debian,a=oldstable,n=jessie,l=Debian,c=main
     origin httpredir.debian.org
Pinned packages:
```
so it seems that name (`n=`) is provided for all of them.

yarikoptic commented on tools/nd_freeze_helper.pl:
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/perl
+# Simple web client for retrieving neurodebian snapshot information.
+# Code adapted from: https://www.oreilly.com/openbook/webclient/ch04.html
```
oy... what was the license for original code if it was used largely to produce this one?
Note that a little disadvantage of a separate script is that users might need to download two scripts now.
BUT cool that we can do that. Also later I might ask to add an option so it doesn't parse anything and just dumps output into a specified file? then I could use it instead of wget in other places

yarikoptic commented on tools/nd_freeze:
```diff
@@ -0,0 +1,334 @@
+# Function to run apt-get update
+sub run_apt_get_update {
+    print "Refreshing apt cache ...\n";
+    qx!apt-get --allow-original-apt-sources update 2> /dev/null!;
```
where did you find this option for `apt-get`?  or is that supposed to be an option for this tool?

yarikoptic replied:
and of cause with `2> /dev/null` no error is spit out ;)  better not to hide those that well!

yarikoptic commented:
also I wonder if it is worth adding a check if "date" requirement could actually be fulfilled, and if not - exit with non-0:

```shell
$> docker run -it --rm -v $PWD/tools:/home ndfreeze:neurodebian-jessie
root@01193a4aee8a:/# eatmydata /home/nd_freeze 20201101 ; echo $?
Discovering installed repository sources ...
Refreshing apt cache ...
0

root@01193a4aee8a:/# cat /etc/apt/sources.list 
# deb http://deb.debian.org/debian jessie main
# deb http://deb.debian.org/debian jessie-updates main
deb http://security.debian.org jessie/updates main

root@01193a4aee8a:/# cat /etc/apt/sources.list.d/snapshots.sources.list 
deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20181011T050502Z// jessie main
deb http://snapshot.debian.org/archive/debian/20181011T155709Z// jessie-updates main
deb http://snapshot.debian.org/archive/debian/20181011T155709Z// jessie main
deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20181011T050502Z// data main
```
In above I specified an impossible date in the future, but I guess snapshot archives bounced to the last known and script proceeded.  Is there an easy way to compare the dates in perl so that the provided date is before/at the date of the snapshot? (or am I missing a usecase where it would be legit to have snapshot date before the requested one if we always go for "the next"?)

yarikoptic commented on tools/tests/test_nd_freeze:
```diff
@@ -0,0 +1,169 @@
+    date=$3
+    keep_sources=${4:-no}
+
+    if [ "$keep_sources" = "no" ]; then
```
since no "standard" test runner is used, it would be nice to add here some printout about what is going on, since otherwise it just sits quietly at "Test basic operation ..." for a while

yarikoptic commented on tools/tests/test_nd_freeze:
```diff
@@ -0,0 +1,169 @@
+    fi
+
+    if grep -Fq$match_param "$line" "$TMP_DIR/$file"; then
+        echo "SUCCESS: line found"
```
May be add which line in what file?  otherwise it is boring list of "SUCCESSes" without a detail to differ them

yarikoptic commented on tools/tests/test_nd_freeze:
```diff
@@ -0,0 +1,169 @@
+        apt-get update
+        apt-get install -y wget
+        wget -O- http://neuro.debian.net/lists/${release}.us-nh.full | tee /etc/apt/sources.list.d/neurodebian.sources.list
+        apt-key adv --recv-keys --keyserver hkp://pool.sks-keyservers.net:80 0xA5D32F012649A5A9
```
why not just use the neurodebian images to avoid all of the above? we should have them for each of debian/ubuntu releases: https://hub.docker.com/_/neurodebian/

yarikoptic commented on tools/tests/test_nd_freeze:
```diff
@@ -0,0 +1,169 @@
+#!/bin/bash
+
+# Temporary scratch directory to save generated files.
+TMP_DIR="tmp_$((1 + RANDOM % 10000))"
```
Why not just `mktemp -d`?

I do see that you remove it in `test_teardown`, and I guess we will take the fact that this directory would remain present if tests fail as a feature, but just FYI you wanted to clean it up upon completion regardless of what has happaned, then you could add line below this line (but before running anything else so it doesn't puke before establishing it):
```shell
trap "rm -rf \"$TMP_DIR\"" 0
```

But since it is not cleaned up and might contain relevant info, better to inform user about which TMP_DIR we are using here (add some `echo`).

ATM I cannot even find this TMP_DIR anywhere upon my crashed attempt to run tests

yarikoptic commented on tools/tests/test_nd_freeze:
```diff
@@ -0,0 +1,169 @@
+    else
+        echo "FAIL: File $file did not contain: $line"
+        test_teardown
+        exit 1
```
I wonder if we should better collect all the failures and report them and then exit with non-0 if any has failed... may be something like `declare -g exit_code; exit_code=0` at the top and then using/assigning it here would suffice to then at the end just `exit $exit_code`?

yarikoptic commented:
@mjtravers pushed two little commits address two little convenience issues to get more diagnostics for the failing test.

yarikoptic commented on tools/tests/test_nd_freeze:
```diff
@@ -0,0 +1,176 @@
+        echo "SUCCESS: line found"
+    else
+        echo -n "FAIL: File $file did not contain: $line. "
+        if [ -e "$file" ]; then
```
I introduced a bug here since `$file` is not full path... fixing that

yarikoptic commented:
ok, here is details for my failure:
```shell
2 13981.....................................:Tue 23 Oct 2018 11:09:30 AM EDT:.
(git)smaug:~exppsy/neurodebian[enh-nd-freeze]tools/tests
$> ./test_nd_freeze
INFO: TMP_DIR=tmp_1021
Test basic operation ...
SUCCESS: line '# deb http://neuro.debian.net/debian data main contrib non-free' found
SUCCESS: line '# deb http://neuro.debian.net/debian data main contrib non-free' found
SUCCESS: line 'deb http://neuro.debian.net/debian data main contrib non-free' found
SUCCESS: line 'deb http://neuro.debian.net/debian data main contrib non-free' found
SUCCESS: line 'deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ jessie non-free contrib main' found
SUCCESS: line 'deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie-updates main' found
SUCCESS: line 'deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ data non-free contrib main' found
SUCCESS: line 'deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie main' found
FAIL: File tmp_1021/snapshots.sources.list did not contain: deb http://snapshot.debian.org/archive/debian-security/20170727T203455Z/ jessie/updates main. Full file content:
deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie main
deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie-updates main
deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ data non-free contrib main
deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ jessie non-free contrib main
chrispycheng has logged on pts/15 from 10.31.113.250.
2 13982 ->1.....................................:Tue 23 Oct 2018 11:24:41 AM EDT:.
```
so it is `jessie-updates` not `jessie/updates`.  But before "fixing it", I wonder @mjtravers - did it pass for you?  (will push my code fix now)

yarikoptic commented:
Also I wonder why my docker is so slow (I guess there is nothing else to explain why it takes awhile for this to run, although it just does `apt-get update`)

mjtravers commented:
@yarikoptic  Latest commit addresses the issues above:

1. Creating the TMP_DIR and cleaning it up
2. Not stopping on FAIL and returning the appropriate exit code
3. Using neurodebian docker images to speed things up
4. Fixing a --no-allow-insecure-repositories switch bug
5. Being able to run the test from any directory
6. Displaying nd_freeze output during the test

yarikoptic commented:
cool, before I try again -- does it pass the test(s) for you? how long does it take?

mjtravers commented:
Yes, the tests are running for me. It completes in 4min 15sec. I have an i7 laptop with 8GBs of RAM.

yarikoptic commented:
eh, I am not that lucky :-(
```shell
(git)smaug:~exppsy/neurodebian[enh-nd-freeze]tools/tests
$> ./test_nd_freeze
[ Test basic operation ]
INFO: TMP_DIR=/home/yoh/.tmp/tmp.KVrTIERKdW
I: Refreshing apt cache ...
I: Discovering installed repository sources ...
I: Refreshing apt cache ...
SUCCESS: line '# deb http://neuro.debian.net/debian jessie main' found
SUCCESS: line '# deb http://neuro.debian.net/debian data main' found
SUCCESS: line 'deb http://neuro.debian.net/debian jessie main' found
SUCCESS: line 'deb http://neuro.debian.net/debian data main' found
SUCCESS: line 'deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ jessie main' found
SUCCESS: line 'deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie-updates main' found
SUCCESS: line 'deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ data main' found
SUCCESS: line 'deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie main' found
FAIL: File /home/yoh/.tmp/tmp.KVrTIERKdW/snapshots.sources.list did not contain: deb http://snapshot.debian.org/archive/debian-security/20170727T203455Z/ jessie/updates main. Full file content:
deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ data main
deb http://snapshot-neuro.debian.net:5002/archive/neurodebian/20170727T050508Z/ jessie main
deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie-updates main
deb http://snapshot.debian.org/archive/debian/20170727T040550Z/ jessie main
SUCCESS: line '# deb http://deb.debian.org/debian jessie main' found
FAIL: File /home/yoh/.tmp/tmp.KVrTIERKdW/sources.list did not contain: # deb http://security.debian.org/debian-security jessie/updates main. Full file content:
# deb http://deb.debian.org/debian jessie main
# deb http://deb.debian.org/debian jessie-updates main
deb http://security.debian.org jessie/updates main
SUCCESS: line '# deb http://deb.debian.org/debian jessie-updates main' found
SUCCESS: line 'deb http://deb.debian.org/debian jessie main' found
FAIL: File /home/yoh/.tmp/tmp.KVrTIERKdW/sources.list.original did not contain: deb http://security.debian.org/debian-security jessie/updates main. Full file content:
deb http://deb.debian.org/debian jessie main
deb http://deb.debian.org/debian jessie-updates main
...
```
so  - "jessie-updates" or "jessie/updates" for debian-security?

mjtravers commented:
Interesting. I'm glad you upgraded the test output.

Your /etc/apt/sources.list for the docker image neurodebian:jessie is different from mine... which can be traced to the "FROM debian:jessie" in the Dockerfile: https://github.com/neurodebian/dockerfiles/blob/88af27888a5aee76924c3b6840fc47f9cd8f63cb/dockerfiles/jessie/Dockerfile

My sources.list is:

deb http://deb.debian.org/debian jessie main
deb http://security.debian.org/debian-security jessie/updates main
deb http://deb.debian.org/debian **jessie-updates** main

This is my docker image digest:

"neurodebian@sha256:e115bf6d8ad4840c17bcff0ac7c8a91146dd37a6b3ecbc55c1eb301e4cf8b9c7"

What's the digest you have?

mjtravers commented:
Ah, there is a problem in the looping in the disable_line() function. Let me fix.

yarikoptic commented:
current failure for me (the tip is):
```
INFO: removing TMP_DIR: /home/yoh/.tmp/tmp.9gAyDQe5wc
[ Test handling of a different release ]
INFO: TMP_DIR=/home/yoh/.tmp/tmp.4TZm1inBdd
grep: /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list: No such file or directory
FAIL: File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list did not contain: # deb http://neuro.debian.net/debian wheezy main. File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list does not exist.
grep: /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list: No such file or directory
FAIL: File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list did not contain: # deb http://neuro.debian.net/debian data main. File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list does not exist.
grep: /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list.original: No such file or directory
FAIL: File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list.original did not contain: deb http://neuro.debian.net/debian wheezy main. File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list.original does not exist.
grep: /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list.original: No such file or directory
FAIL: File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list.original did not contain: deb http://neuro.debian.net/debian data main. File /home/yoh/.tmp/tmp.4TZm1inBdd/neurodebian.sources.list.original does not exist.
grep: /home/yoh/.tmp/tmp.4TZm1inBdd/snapshots.sources.list: No such file or directory
...
```

yarikoptic commented:
@mjtravers I am still fighting the strange crash of `docker run` in the 2nd test.  Part of the problem is that nd_freeze just exits without anything being printed by `nd_freeze` whatsoever so it is hard-to-impossible to figure out what is going on wrong:
```shell
$> docker run -it --rm -v /home/yoh/deb/gits/pkg-exppsy/neurodebian/tools/tests/./..:/tests -v /home/yoh/.tmp/tmp.F7tdDPdh16:/temp neurodebian:wheezy /bin/bash -c '
        /tests/nd_freeze  1/2/2018
    '
$> echo $?
139
```

I have pushed 8bd3dc9363ab837c01252e91eecf808081f49583  with few enhancements (and a fix to detect docker failure in the pipe).
May be `nd_freeze` could get `--debug` flag which would trigger additional debug messages along the way?

yarikoptic commented on tools/nd_freeze:
```diff
@@ -0,0 +1,391 @@
+sub run_apt_get_update {
+    info "Refreshing apt cache";
+    if (qx!grep ubuntu /etc/os-release!) {
+        print qx/apt-get update --no-allow-insecure-repositories/;
```
Could those print become a debug function similar to info? So later on we could add option to enable or disable them

yarikoptic commented:
grr... even with print outs, the one which fails, fails silently (so it is before those commands output of which you started to print):
```shell
$> ./test_nd_freeze
[ Test basic operation ]
INFO: TMP_DIR=/home/yoh/.tmp/tmp.gZwpMIoi7I distro=neurodebian release=jessie date=7/27/2017 keep_sources=no
jessie: Pulling from library/neurodebian
Digest: sha256:e115bf6d8ad4840c17bcff0ac7c8a91146dd37a6b3ecbc55c1eb301e4cf8b9c7
Status: Image is up to date for neurodebian:jessie
INFO: Refreshing apt cache
Get:1 http://neuro.debian.net jessie InRelease [24.0 kB]
...
[ Test handling of a different release ]
INFO: TMP_DIR=/home/yoh/.tmp/tmp.Jslt8Cneha distro=neurodebian release=wheezy date=1/2/2018 keep_sources=no
wheezy: Pulling from library/neurodebian
Digest: sha256:6b43526442dcf90fba95cb367af95ff991edae04a7b66f3eff8bb9042561bea3
Status: Image is up to date for neurodebian:wheezy
ERROR: docker run failed.  Exit code was 139
Full command was: docker run -it --rm -v /home/yoh/deb/gits/pkg-exppsy/neurodebian/tools/tests/./..:/tests -v /home/yoh/.tmp/tmp.Jslt8Cneha:/temp neurodebian:wheezy /bin/bash -c '
        /tests/nd_freeze  1/2/2018
        if [ -f /etc/apt/sources.list ]; then
            cp /etc/apt/sources.list /temp/sources.list
        fi
        if [ -f /etc/apt/sources.list.original ]; then
            cp /etc/apt/sources.list.original /temp/sources.list.original
        fi
        if [ -f /etc/apt/sources.list.d/neurodebian.sources.list ]; then
            cp /etc/apt/sources.list.d/neurodebian.sources.list /temp/neurodebian.sources.list
        fi
        if [ -f /etc/apt/sources.list.d/neurodebian.sources.list.original ]; then
            cp /etc/apt/sources.list.d/neurodebian.sources.list.original /temp/neurodebian.sources.list.original
        fi
        if [ -f /etc/apt/sources.list.d/snapshots.sources.list ]; then
            cp /etc/apt/sources.list.d/snapshots.sources.list /temp/snapshots.sources.list
        fi
    '
```
why am I special again!?!?!

yarikoptic commented:
uff... I should have suspected that
```shell
(git)smaug:~exppsy/neurodebian[enh-nd-freeze]tools/tests
$> docker run -it --rm debian:wheezy /bin/dash
# bash                                        
Segmentation fault (core dumped)
# uname
Linux
# uname -a
Linux 796794c8bd7c 4.16.0-0.bpo.2-amd64 #1 SMP Debian 4.16.16-2~bpo9+1 (2018-06-26) x86_64 GNU/Linux
```
not sure yet on how to deal with it anyhow.
merging, thanks!
